factor/totp/classes/factor.php

```diff
@@ -354,6 +355,16 @@ public function setup_user_factor($data) {
             $row->lastverified = 0;
             $row->revoked = 0;
 
+            // Check if a record with this secret already exists for the same user. If so, throw an exception.
+            $exists = $DB->record_exists('tool_mfa', [
+                'userid' => $row->userid,
+                'secret' => $row->secret,
+                'factor' => $row->factor,
+            ]);
+            if ($exists) {
+                throw new \moodle_exception('error:alreadyregistered', 'factor_totp');
```

Comment:
Could this just be an alert notification and instead of an exception? Little less disruptive to the user

Reply:
Yes, I'll change this.

Reply:
So follow up question. Should `setup_user_factor` return the existing record? Or `null` in this case?

If it returns null, then it throws a different error `Can not setup factor` and we end up in the same place, with a bit more noise.

So I've opted to return the record instead, and set the notification to a warning so the user is redirected to the list of factors instead of back to the TOTP setup page.

Please let me know what you expect them to land on if it already exists.